README.md
<div id="top"></div>

<h3 align="center">siyuanhelper</h3>
  <p align="center">
    The Helper Lib for Siyuan Note.
    <br />
    <a href="https://clouder0.github.io/siyuanhelper/"><strong>Explore the docs »</strong></a>
  </p>
</div>

## 📜 TOC

<details><summary>Table of Contents</summary>

- [🌟 Badges](#🌟-badges)
- [💡 Introduction](#💡-introduction)
- [✨ Features](#✨-features)
- [🎏 Getting Started](#🎏-getting-started)
- [🗺️ Roadmap](#🗺️-roadmap)
- [❓ Faq](#❓-faq)
- [💌 Contributing](#💌-contributing)
- [🙏 Acknowledgment](#🙏-acknowledgment)
- [📖 License](#📖-license)
- [📧 Contact](#📧-contact)

</details>

## 🌟 Badges

[![Test][github-action-test-shield]][github-action-test-url]
[![Codecov][codecov-shield]][codecov-url]
[![pre-commit-ci][pre-commit-ci-shield]][pre-commit-ci-url]
[![pepy-shield]][pepy-url]

[![release-shield]][release-url]
[![pyversions-shield]][pyversions-url]
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![CodeFactor-shield]][CodeFactor-url]
[![code-style-black-shield]][code-style-black-url]

<!-- INTRODUCTION -->
## 💡 Introduction

Siyuan Helper is a python API wrapper for [Siyuan Note](https://github.com/siyuan-note/siyuan).

<p align="right">(<a href="#top">back to top</a>)</p>

## ✨ Features

- Aysnc
- Object Oriented
- Type Annotated
- Fully Documented

<p align="right">(<a href="#top">back to top</a>)</p>

## 🎏 Getting Started

See our [Quickstart](https://clouder0.github.io/siyuanhelper/quickstart/) to get hands on experience.

<p align="right">(<a href="#top">back to top</a>)</p>

## 🗺️ Roadmap

Please check out our [Github Project](https://github.com/Clouder0/siyuanhelper/projects/1).

See the [open issues](https://github.com/Clouder0/siyuanhelper/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#top">back to top</a>)</p>

## ❓ FAQ

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- CONTRIBUTING -->
## 💌 Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

Don't forget to see our [Contributing Guidelines](https://github.com/Clouder0/siyuanhelper/blob/main/CONTRIBUTING.md) for details.

<p align="right">(<a href="#top">back to top</a>)</p>

## 🙏 Acknowledgment

There are various open source projects that siyuanhelper depends on, without which this tool wouldn't exist. Credits to them!

- [aiohttp](https://github.com/aio-libs/aiohttp), Apache License 2.0

<p align="right">(<a href="#top">back to top</a>)</p>

## 📖 License

Distributed under the MIT License. See `LICENSE` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>

## 📧 Contact

Clouder0's email: clouder0@outlook.com

Project Link: [https://github.com/Clouder0/siyuanhelper](https://github.com/Clouder0/siyuanhelper)

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/Clouder0/siyuanhelper.svg?style=for-the-badge
[contributors-url]: https://github.com/Clouder0/siyuanhelper/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/Clouder0/siyuanhelper.svg?style=for-the-badge
[forks-url]: https://github.com/Clouder0/siyuanhelper/network/members
[stars-shield]: https://img.shields.io/github/stars/Clouder0/siyuanhelper.svg?style=for-the-badge
[stars-url]: https://github.com/Clouder0/siyuanhelper/stargazers
[issues-shield]: https://img.shields.io/github/issues/Clouder0/siyuanhelper.svg?style=for-the-badge
[issues-url]: https://github.com/Clouder0/siyuanhelper/issues
[license-shield]: https://img.shields.io/github/license/Clouder0/siyuanhelper.svg?style=for-the-badge
[license-url]: https://github.com/Clouder0/siyuanhelper/blob/main/LICENSE
[github-action-test-shield]: https://github.com/Clouder0/siyuanhelper/actions/workflows/test.yml/badge.svg?branch=main
[github-action-test-url]: https://github.com/Clouder0/siyuanhelper/actions/workflows/test.yml
[codecov-shield]:https://codecov.io/gh/Clouder0/siyuanhelper/branch/main/graph/badge.svg?token=D2XT099AFB
[codecov-url]: https://codecov.io/gh/Clouder0/siyuanhelper
[pre-commit-ci-shield]: https://results.pre-commit.ci/badge/github/Clouder0/siyuanhelper/main.svg
[pre-commit-ci-url]: https://results.pre-commit.ci/latest/github/Clouder0/siyuanhelper/main
[code-style-black-shield]: https://img.shields.io/badge/code%20style-black-000000.svg?style=for-the-badge
[code-style-black-url]: https://github.com/psf/black
[pyversions-shield]: https://img.shields.io/pypi/pyversions/siyuanhelper.svg?style=for-the-badge
[pyversions-url]: https://pypi.org/project/siyuanhelper/
[release-shield]: https://img.shields.io/github/release/Clouder0/siyuanhelper.svg?style=for-the-badge
[release-url]: https://github.com/Clouder0/siyuanhelper/releases
[CodeFactor-shield]: https://www.codefactor.io/repository/github/clouder0/siyuanhelper/badge/main?style=for-the-badge
[CodeFactor-url]: https://www.codefactor.io/repository/github/clouder0/siyuanhelper/overview/main
[pepy-shield]: https://static.pepy.tech/personalized-badge/siyuanhelper?period=total&units=international_system&left_color=grey&right_color=brightgreen&left_text=Downloads
[pepy-url]: https://pepy.tech/project/siyuanhelper


CHANGELOG.md
# Change Log

See [keep-a-changelog](https://github.com/olivierlacan/keep-a-changelog).

Version numbers follow [semantic versioning](https://semver.org).

## Unreleased

## v0.3.0

### Added

- support default value in BlockAttr.get()
- add tags() in SiyuanBlock and tag in RawSiyuanBlock
- cached property
- add create_doc_with_md
- add get_hpath_by_id/path
- add remove_doc
- add parent & son suuport

## v0.2.0

### Added

- Export Markdown Content

## v0.1.0

Huge refactor. Brand new start.

### Added

- Siyuan API Class
- SiyuanBlock Class
- BlockAttr Class
- Document using Mkdocs and Mkdocs-Material
- Development
  - Poetry for Managing Project
  - Mypy for Type Annotations
  - Pre-Commit Check
  - Tox Autonamted Testing
  - Pytest
  - Flake8
  - Black for Code Style
  - isort for Import Sorting

## TEMPLATE

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security


tests/test_api.py
from __future__ import annotations

import asyncio

import pytest
import siyuanhelper.api

from siyuanhelper import exceptions
from siyuanhelper.api import DataType, RawSiyuanBlock, Siyuan, SiyuanBlock


@pytest.fixture
async def siyuan():
    my_siyuan = Siyuan(token="mhfj23ilil7bafg0")
    yield my_siyuan
    await my_siyuan.close()


class TestSiyuan:
    @pytest.mark.asyncio_cooperative
    async def test_get_block_by_id(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn")
        assert ret.asdict() == {
            "alias": "",
            "box": "20220501134144-oqwd5yh",
            "content": "test block",
            "created": "20220501134149",
            "fcontent": "",
            "hash": "6f7a904",
            "hpath": "/test",
            "ial": '{: id="20220501134149-qexauwn" updated="20220501134155"}',
            "id": "20220501134149-qexauwn",
            "length": 10,
            "markdown": "test block",
            "memo": "",
            "name": "",
            "parent_id": "20220501134149-eafvjwu",
            "path": "/20220501134149-eafvjwu.sy",
            "root_id": "20220501134149-eafvjwu",
            "sort": 10,
            "subtype": "",
            "type": "p",
            "updated": "20220501134155",
            "tag": "",
        }

    @pytest.mark.asyncio_cooperative
    async def test_get_raw_block_by_id(self, siyuan: Siyuan):
        ret = await siyuan._get_raw_block_by_id("20220501134149-qexauwn")
        assert ret.id == "20220501134149-qexauwn"

    @pytest.mark.asyncio_cooperative
    async def test_get_attrs_by_id(self, siyuan: Siyuan):
        ret = await siyuan.get_attrs_by_id("20220501214630-ql8hhto")
        assert ret["custom-testattr"] == "value"

    @pytest.mark.benchmark
    @pytest.mark.longrun
    def test_benchmark_raw(self, benchmark):
        ret = {
            "alias": "",
            "box": "20220501134144-oqwd5yh",
            "content": "test block",
            "created": "20220501134149",
            "fcontent": "",
            "hash": "6f7a904",
            "hpath": "/test",
            "ial": '{: id="20220501134149-qexauwn" updated="20220501134155"}',
            "id": "20220501134149-qexauwn",
            "length": 10,
            "markdown": "test block",
            "memo": "",
            "name": "",
            "parent_id": "20220501134149-eafvjwu",
            "path": "/20220501134149-eafvjwu.sy",
            "root_id": "20220501134149-eafvjwu",
            "sort": 10,
            "subtype": "",
            "type": "p",
            "updated": "20220501134155",
            "tag": "",
        }

        def init_raw():
            RawSiyuanBlock(**ret)

        benchmark(init_raw)

    @pytest.mark.benchmark
    @pytest.mark.longrun
    def test_benchmark_filtered_raw(self, benchmark):
        ret = {
            "alias": "",
            "box": "20220501134144-oqwd5yh",
            "content": "test block",
            "created": "20220501134149",
            "fcontent": "",
            "hash": "6f7a904",
            "hpath": "/test",
            "ial": '{: id="20220501134149-qexauwn" updated="20220501134155"}',
            "id": "20220501134149-qexauwn",
            "length": 10,
            "markdown": "test block",
            "memo": "",
            "name": "",
            "parent_id": "20220501134149-eafvjwu",
            "path": "/20220501134149-eafvjwu.sy",
            "root_id": "20220501134149-eafvjwu",
            "sort": 10,
            "subtype": "",
            "type": "p",
            "updated": "20220501134155",
            "tag": "",
        }

        def init_filtered_raw():
            RawSiyuanBlock(**{key: ret[key] for key in siyuanhelper.api.block_fields})

        benchmark(init_filtered_raw)

    @pytest.mark.asyncio_cooperative
    async def test_auth(self):
        new_siyuan = Siyuan(token="invalid")
        with pytest.raises(exceptions.SiyuanAuthFailedException):
            await new_siyuan.sql_query("select * from blocks limit 0")

    @pytest.mark.asyncio_cooperative
    async def test_get_blocks_by_sql(self, siyuan: Siyuan):
        blocks: list[SiyuanBlock] = await siyuan.get_blocks_by_sql(
            "WHERE (id='20220501134149-qexauwn') or (id='20220501134156-v82db6e')"
        )
        assert (
            blocks[0].id == "20220501134149-qexauwn"
            and blocks[1].id == "20220501134156-v82db6e"
        ) or (
            blocks[1].id == "20220501134149-qexauwn"
            and blocks[0].id == "20220501134156-v82db6e"
        )

    @pytest.mark.asyncio_cooperative
    async def test_create_remove_document(self, siyuan: Siyuan):
        block = await siyuan.create_doc_with_md(
            "20220501134144-oqwd5yh", "/testfolder/testdoc", "test content\n\ntest2"
        )
        await asyncio.sleep(4)
        await block.ensure()
        content = await block.export()
        assert content == "test content\n\ntest2\n"
        await siyuan.remove_doc("20220501134144-oqwd5yh", block.path)
        await asyncio.sleep(4)
        with pytest.raises(exceptions.SiyuanApiException):
            await siyuan.get_hpath_by_id(block.id)

    @pytest.mark.asyncio_cooperative
    async def test_get_doc_path_by_id(self, siyuan: Siyuan):
        doc = await siyuan.get_doc_path_by_id("20220728223910-bkzr0sb")
        assert doc["box"] == "20220501134144-oqwd5yh"
        assert doc["path"] == "/20220728220136-wi8vton/20220728223910-bkzr0sb.sy"

    @pytest.mark.asyncio_cooperative
    async def test_get_hpath_by_id(self, siyuan: Siyuan):
        hpath = await siyuan.get_hpath_by_id("20220728223630-00qikyt")
        assert hpath == "/testfolder/testhpathbyid"

    @pytest.mark.asyncio_cooperative
    async def test_get_hpath_by_path(self, siyuan: Siyuan):
        hpath = await siyuan.get_hpath_by_path(
            "20220501134144-oqwd5yh",
            "/20220728220136-wi8vton/20220728223910-bkzr0sb.sy",
        )
        assert hpath == "/testfolder/hpath by path"

    @pytest.mark.asyncio_cooperative
    async def test_get_parent_id_by_id(self, siyuan: Siyuan):
        parent_id = await siyuan.get_parent_id_by_id("20220501134149-qexauwn")
        assert parent_id == "20220501134149-eafvjwu"

    # @pytest.mark.asyncio_cooperative
    # async def test_rename_doc(self, siyuan: Siyuan):
    #     doc = await siyuan.get_doc_path_by_id("20220808150136-2h756le")
    #     rd = str(random.randint(0, 100))
    #     await siyuan.rename_doc(doc["box"], doc["path"], "renamed" + rd)
    #     await asyncio.sleep(5)
    #     content = await siyuan.sql_query(
    #         f"select content from blocks where path='{doc['path']}'"
    #     )
    #     assert content[0]["content"] == "renamed" + rd

    # @pytest.mark.asyncio_cooperative
    # async def test_move_doc(self, siyuan: Siyuan):
    #     await siyuan.move_doc(
    #         "20220501134144-oqwd5yh",
    #         "/20220808153645-lzljvgw.sy",
    #         "20220501134144-oqwd5yh",
    #         "/20220808153741-sf7rb5j/",
    #     )
    #     await asyncio.sleep(5)
    #     doc = await siyuan.get_doc_path_by_id("20220808153645-lzljvgw")
    #     assert doc["path"] == "/20220808153741-sf7rb5j/20220808153645-lzljvgw.sy"
    #     await siyuan.move_doc(
    #         "20220501134144-oqwd5yh",
    #         "/20220808153741-sf7rb5j/20220808153645-lzljvgw.sy",
    #         "20220501134144-oqwd5yh",
    #         "/",
    #     )


class TestSiyuanBlock:
    @pytest.mark.asyncio_cooperative
    async def test_proxy_getattr(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn")
        assert ret.hash == "6f7a904"

    @pytest.mark.asyncio_cooperative
    async def test_invalid_attr(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn")
        assert ret.test is None

    @pytest.mark.asyncio_cooperative
    async def test_pull(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn", full=False)
        assert ret.raw is None
        await ret.pull()
        assert ret.content == "test block"

    @pytest.mark.longrun
    @pytest.mark.asyncio_cooperative
    async def test_insert_delete(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn")
        inserted = await ret.insert(DataType.MARKDOWN, "test insert")
        await asyncio.sleep(5)
        await inserted.ensure()
        assert inserted.markdown == "test insert"
        await inserted.delete()
        await asyncio.sleep(5)
        with pytest.raises(exceptions.SiyuanNoResultException):
            await siyuan.get_block_by_id(inserted.id)

    @pytest.mark.asyncio_cooperative
    async def test_export(self, siyuan: Siyuan):
        block = await siyuan.get_block_by_id("20220507183011-c3lt1k8")
        ret = await block.export()
        assert (
            ret
            == "**content1**\n\n*content2*\n\n## Header 2\n\ncontent 3\n\ncontent 4\n"
        )

    @pytest.mark.asyncio_cooperative
    async def test_tags(self, siyuan: Siyuan):
        block = await siyuan.get_block_by_id("20220501231358-j4mnnht")
        assert (await block.tags) == ("tag", "tag2")
        assert (await block.tags) == ("tag", "tag2")  # test cache


class TestBlockAttr:
    @pytest.mark.asyncio_cooperative
    async def test_attr_get(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501214630-ql8hhto")
        assert await ret.attrs.get("custom-testattr") == "value"

    @pytest.mark.asyncio_cooperative
    async def test_attr_set(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501214630-ql8hhto")
        assert await ret.attrs.get("custom-testattr") == "value"
        assert await ret.attrs.get("custom-testattr2") == ""
        await ret.attrs.set("custom-testattr2", "value2")
        assert await ret.attrs.get("custom-testattr2") == "value2"
        await ret.attrs.set("custom-testattr2", "")
        assert await ret.attrs.get("custom-testattr2") == ""

    @pytest.mark.asyncio_cooperative
    async def test_cache(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501214630-ql8hhto")
        assert (await ret.attrs.values)["custom-testattr"] == "value"
        assert (await ret.attrs.values)["custom-testattr"] == "value"

    @pytest.mark.asyncio_cooperative
    async def test_attr_get_default(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501214630-ql8hhto")
        assert await ret.attrs.get("custom-testattr-none") == ""
        assert await ret.attrs.get("custom-testattr-none", "default") == "default"

    @pytest.mark.asyncio_cooperative
    async def test_get_parent(self, siyuan: Siyuan):
        ret = await siyuan.get_block_by_id("20220501134149-qexauwn")
        pa = await ret.parent
        assert pa.id == "20220501134149-eafvjwu"

    @pytest.mark.asyncio_cooperative
    async def test_get_sons(self, siyuan: Siyuan):
        block = await siyuan.get_block_by_id("20220728223628-en9be7r")
        sons = await block.sons
        son_ids = [x.id for x in sons]
        assert "20220909163636-mr190sz" in son_ids
        assert "20220909163636-q2maalx" in son_ids


src/siyuanhelper/api.py
"""main API module."""
from __future__ import annotations

import dataclasses

from dataclasses import dataclass
from enum import Enum
from typing import Any, Union, cast

import aiohttp

from asyncstdlib import cached_property
from siyuanhelper import exceptions


data_type = Union[dict, list, None]


class Siyuan:
    """Siyuan Helper Instance."""

    def __init__(self, base_url: str = "http://127.0.0.1:6806", token: str = ""):
        """Init a Siyuan Helper.

        Args:
            base_url (str, optional): the url to invoke requests. Defaults to "http://127.0.0.1:6806".
            token (str, optional): API token, none if unused. Defaults to "".

        Raises:
            exceptions.SiyuanAuthFailedException: raised if Authorization Failed.
        """
        self.base_url = base_url
        self.token = token
        self.session = aiohttp.ClientSession(
            self.base_url,
            headers={
                "Authorization": f"Token {token}",
                "Content-Type": "application/json",
            },
        )

    async def close(self) -> None:
        """Close Siyuan Helper Session, should be explicitly called after use."""
        await self.session.close()

    async def _post(self, url: str, **params: Any) -> data_type:
        async with self.session.post(url=url, json=params) as resp:
            ret = SiyuanResponse(**(await resp.json()))
            if ret.code == 0:
                return ret.data
            if ret.code == -1 and ret.msg == "Auth failed":
                raise exceptions.SiyuanAuthFailedException((self, ret))
            else:
                raise exceptions.SiyuanApiException((self, ret))

    async def get_block_by_id(self, block_id: str, full: bool = True) -> SiyuanBlock:
        """Get SiyuanBlock by block id.

        Args:
            block_id (str): the desired block id.
            full (bool): whether to fetch all the informations. Defaults to True.

        Returns:
            SiyuanBlock: the block with all fields.
        """
        if not full:
            return SiyuanBlock(id=block_id, source=self)
        return SiyuanBlock(
            id=block_id, source=self, raw=await self._get_raw_block_by_id(block_id)
        )

    async def get_blocks_by_sql(
        self, cond: str, full: bool = True
    ) -> tuple[SiyuanBlock, ...]:
        """Get a list of SiyuanBlock by sql.

        Args:
            cond (str): the conditions to apply, typically `where id = ''` or so.
            full (bool, optional): whether to fetch all the informations of the block. Defaults to True.

        Returns:
            tuple[SiyuanBlock, ...]: result blocks
        """
        if not full:
            ret = await self.sql_query(f"SELECT id from BLOCKS {cond}")
            if ret is None:
                return ()
            return tuple(SiyuanBlock(id=x["id"], source=self) for x in ret)
        ret = await self.sql_query(f"SELECT * from BLOCKS {cond}")
        if ret is None:
            return ()
        return tuple(
            SiyuanBlock(id=x["id"], source=self, raw=self._gen_block_by_sql_result(x))
            for x in ret
        )

    def _gen_block_by_sql_result(self, result: dict) -> RawSiyuanBlock:
        # use block_fields filter to avoid compatibility issues.
        return RawSiyuanBlock(**{key: result[key] for key in block_fields})

    async def _get_raw_block_by_id(self, block_id: str) -> RawSiyuanBlock:
        """Generally, you should not use this function unless you know what you're doing. Get RawSiyuanBlock by block id.

        Args:
            block_id (str): the desired block id.

        Returns:
            RawSiyuanBlock: raw Siyuan Block, with only data fields defined.
        """
        ret = await self.sql_query(f"SELECT * from BLOCKS where ID = '{block_id}'")
        if type(ret) != list:
            raise exceptions.SiyuanApiTypeException(ret)
        if len(ret) == 0:
            raise exceptions.SiyuanNoResultException(ret)
        return self._gen_block_by_sql_result(ret[0])

    async def get_attrs_by_id(self, block_id: str) -> dict[str, str]:
        """Get attribute dictionary by block id.

        Args:
            block_id (str): target block.

        Returns:
            dict[str, str]: key-value dict, note that custom attributes starts with `custom-`
        """
        ret = await self._post("/api/attr/getBlockAttrs", id=block_id)
        if type(ret) != dict:
            raise exceptions.SiyuanApiTypeException
        return ret

    async def set_attrs_by_id(self, block_id: str, attrs: dict[str, str]) -> None:
        """Update the attributes of the block with given id. Won't delete attrs not given in the dict.

        Args:
            block_id (str): target block id
            attrs (dict[str, str]): block attrs dict to update
        """
        await self._post("/api/attr/setBlockAttrs", id=block_id, attrs=attrs)

    async def sql_query(self, sql: str) -> data_type:
        """Query SQL.

        Args:
            sql (str): the executed SQL string

        Returns:
            data_type: usually a list of dicts.
        """
        return await self._post(url="/api/query/sql", stmt=sql)

    async def get_parent_id_by_id(self, id: str) -> str:
        """Query the Parent Id of a block.

        Args:
            id (str): target block id.

        Returns:
            str: Parent block id.
        """
        return (await self.sql_query(f"SELECT parent_id FROM blocks WHERE id='{id}'"))[
            0
        ]["parent_id"]

    async def delete_block_by_id(self, block_id: str) -> None:
        """Delete a block with given id.

        Args:
            block_id (str): target block id
        """
        await self._post("/api/block/deleteBlock", id=block_id)

    async def insert_block(
        self, data_type: DataType, data: str, previous_id: str
    ) -> SiyuanBlock:
        """Insert a block after the block with the given id.

        Args:
            data_type (DataType): markdown or dom
            data (str): data value
            previous_id (str): the block in front of the new block

        Raises:
            exceptions.SiyuanApiException: API Error

        Returns:
            SiyuanBlock: the new block, with id only.
        """
        ret = await self._post(
            "/api/block/insertBlock",
            dataType=data_type,
            data=data,
            previousID=previous_id,
        )
        if ret is None:
            raise exceptions.SiyuanApiException((self, ret))
        return await self.get_block_by_id(ret[0]["doOperations"][0]["id"], full=False)

    async def export_md_content_by_id(self, block_id: str) -> str:
        """Export Markdown Content by id.

        Args:
            block_id (str): blockid, only document block is supported.

        Returns:
            str: markdown
        """
        return cast(dict, await self._post("/api/export/exportMdContent", id=block_id))[
            "content"
        ]

    async def create_doc_with_md(
        self, notebook_id: str, path: str, markdown: str
    ) -> SiyuanBlock:
        """Create a doc with markdown content.

        Args:
            notebook_id (str): id of the notebook where the doc is to be created
            path (str): path of the doc, eg. '/foo/bar'
            markdown (str): the markdown content of the doc.

        Returns:
            SiyuanBlock: the created document block, with only id field.
        """
        ret = await self._post(
            "/api/filetree/createDocWithMd",
            notebook=notebook_id,
            path=path,
            markdown=markdown,
        )
        return await self.get_block_by_id(cast(str, ret), False)

    async def get_doc_path_by_id(self, doc_id: str) -> dict:
        """Get the notebook_id and raw path from document id.

        Args:
            doc_id (str): the target document id.

        Returns:
            dict: {box: notebook_id, path: raw path}
        """
        ret = cast(
            list,
            (await self.sql_query(f"SELECT box, path FROM blocks WHERE id='{doc_id}'")),
        )[0]
        return ret

    async def get_hpath_by_id(self, doc_id: str) -> str:
        """Get readable path by Document id.

        Args:
            doc_id (str): the id of the document.

        Returns:
            str: human readable path of the document.
        """
        ret = await self._post("/api/filetree/getHPathByID", id=doc_id)
        return cast(str, ret)

    async def get_hpath_by_path(self, notebook_id: str, doc_path: str) -> str:
        """Get doc's readable path by raw path.

        Args:
            notebook_id (str): the id of the notebook where document locates.
            doc_path (str): raw path, eg. '/20220728220136-wi8vton/20220728223910-bkzr0sb.sy'

        Returns:
            str: human readable path.
        """
        ret = await self._post(
            "/api/filetree/getHPathByPath", notebook=notebook_id, path=doc_path
        )
        return cast(str, ret)

    async def remove_doc(self, notebook_id: str, doc_path: str) -> None:
        """Remove the target document.

        Args:
            notebook_id (str): notebook id which the document belongs to.
            doc_path (str): path of the target document.
        """
        await self._post("/api/filetree/removeDoc", notebook=notebook_id, path=doc_path)

    # async def rename_doc(self, notebook_id: str, path: str, new_title: str) -> None:
    #     """Rename the target document.

    #     Args:
    #         notebook_id (str): notebook id where the doc locates.
    #         path (str): the raw path of the target document.
    #         new_title (str): new title of the document.
    #     """
    #     await self._post(
    #         "/api/filetree/renameDoc", notebook=notebook_id, path=path, title=new_title
    #     )

    # async def move_doc(
    #     self, notebook_id1: str, path1: str, notebook_id2: str, path2: str
    # ) -> None:
    #     """Move the target document to a new location.

    #     Args:
    #         notebook_id1 (str): the old document notebook.
    #         path1 (str): old path.
    #         notebook_id2 (str): new notebook id.
    #         path2 (str): new dir path. eg. `"/20220808153741-sf7rb5j/"`
    #     """
    #     await self._post(
    #         "/api/filetree/moveDoc",
    #         fromNotebook=notebook_id1,
    #         fromPath=path1,
    #         toNotebook=notebook_id2,
    #         toPath=path2,
    #     )


@dataclass
class SiyuanResponse:
    """Response class for siyuan."""

    code: int
    msg: str
    data: data_type = None


class BlockAttr:
    """Block Attribute Class."""

    def __init__(self, block: SiyuanBlock):
        """Init.

        Args:
            block (SiyuanBlock): block that this BlockAttr adhere to.
        """
        self.block = block

    @cached_property
    async def values(self) -> dict[str, str]:
        """Return the dictionary of the attributes. Cached.

        Returns:
            dict[str, str]: dict of attributes.
        """
        return await self.block.source.get_attrs_by_id(self.block.id)

    async def get(self, name: str, default: str = "") -> str:
        """Get attribute value by name.

        Args:
            name (str): name of the attribute, remember to add `custom-`
            default (str, optional): the return value if no attribute is found, defaults to ""

        Returns:
            str: the value of the attribute, default if not found.
        """
        return (await self.values).get(name, default)

    async def set(self, name: str, val: str) -> None:
        """Modify the attribute.

        Args:
            name (str): name of the attribute
            val (str): new value
        """
        await self.block.source.set_attrs_by_id(self.block.id, {name: val})
        del self.values


class DataType(str, Enum):
    """DataType Enum, used when modifying block's content."""

    MARKDOWN = "markdown"
    DOM = "dom"


class SiyuanBlock:
    """Block Class for Siyuan. An additional application layer is applied. For raw data, consider RawSiyuanBlock."""

    def __init__(self, id: str, source: Siyuan, raw: RawSiyuanBlock | None = None):
        """Init a SiyuanBlock.

        Args:
            id (str): id of the block.
            source (Siyuan): source of the block.
            raw (RawSiyuanBlock | None, optional): raw block data. Defaults to None.
        """
        self.id = id
        self.source = source
        self.raw = raw
        self.attrs = BlockAttr(self)

    @cached_property
    async def tags(self) -> tuple[str, ...]:
        """Return the tags tuple of the block. Note that if the Block hasn't been loaded, this function will pull the block from API.

        Returns:
            tuple[str]: tag tuple, such as `("tag1", "tag2")`
        """
        await self.ensure()
        return tuple(x.strip("#") for x in self.raw.tag.split(" "))  # type: ignore

    @cached_property
    async def parent(self) -> SiyuanBlock | None:
        """Get the parent block of the current block.

        Returns:
            SiyuanBlock | None: Parent block. None if current block is root.
        """
        pid = await self.source.get_parent_id_by_id(self.id)
        if pid == "":
            return None
        return await self.source.get_block_by_id(pid, False)

    @cached_property
    async def sons(self) -> tuple[SiyuanBlock, ...]:
        """Get a tuple of blocks whose parent is the current block.

        Returns:
            tuple[SiyuanBlock, ...]: sons of the current block.
        """
        return await self.source.get_blocks_by_sql(
            f"WHERE parent_id='{self.id}'", False
        )

    async def pull(self) -> None:
        """Pull from Siyuan API. Refreshing everything."""
        self.raw = await self.source._get_raw_block_by_id(self.id)
        await self.attrs.values

    async def ensure(self) -> None:
        """Ensure the information of the current block is cached."""
        if self.raw is None:
            self.raw = await self.source._get_raw_block_by_id(self.id)

    def asdict(self) -> dict:
        """Parse Siyuan Block to a dict containing all its informations.

        Returns:
            dict: that block.
        """
        return dataclasses.asdict(self.raw)

    def __getattr__(self, __name: str) -> Any:
        """Expose RawSiyuanBlock's attributes.

        Args:
            __name (str): attribute name

        Returns:
            Any: result
        """
        if self.raw is not None and __name in self.raw.__slots__:  # type: ignore
            return self.raw.__getattribute__(__name)

    async def delete(self) -> None:
        """Delete this block. Mind that there is a delay between the execution and the result being synced into API database."""
        await self.source.delete_block_by_id(self.id)

    async def insert(self, data_type: DataType, data: str) -> SiyuanBlock:
        """Insert a block after this block.

        Args:
            data_type (DataType): markdown or dom
            data (str): the desired data

        Returns:
            SiyuanBlock: newly inserted block, only `id` is given.
        """
        return await self.source.insert_block(data_type, data, self.id)

    async def export(self) -> str:
        """Export the document current block belongs to in markdown format.

        Returns:
            str: markdown export output
        """
        return await self.source.export_md_content_by_id(self.id)


block_fields = (
    "id",
    "alias",
    "box",
    "content",
    "created",
    "updated",
    "fcontent",
    "hash",
    "hpath",
    "length",
    "markdown",
    "memo",
    "name",
    "parent_id",
    "path",
    "root_id",
    "sort",
    "subtype",
    "type",
    "ial",
    "tag",
)


@dataclass(frozen=True)
class RawSiyuanBlock:
    """Raw Siyuan Block, presents the raw output of the Siyuan API."""

    __slots__ = block_fields

    id: str
    alias: str
    box: str
    content: str
    created: str
    updated: str
    fcontent: str
    hash: str
    hpath: str
    length: int
    markdown: str
    memo: str
    name: str
    parent_id: str
    path: str
    root_id: str
    sort: int
    subtype: str
    type: str
    ial: str
    tag: str
